exporter/exporterhelper/internal/sized_elements_channel.go

```diff
@@ -0,0 +1,111 @@
+	cap int64
+	ch  chan T
+
+	// The following fields are used by the persistent queue to initialize the queue with the elements recovered from
+	// the disk. They have to be tracked separately because the actual total size of all the restored elements is not
+	// guaranteed to be equal initTotalSize.
+	initTotalSize     int64
+	initElementsCount int64
+	initSizeMutex     sync.Mutex
+}
+
+// newSizedElementsChannel creates a sized elements channel. Each element is assigned a size by the provided sizer.
+// chanCapacity is the capacity of the underlying channel which usually should be equal to the capacity of the queue to
+// avoid blocking the producer.
+func newSizedElementsChannel[T any](capacity int) *sizedElementsChannel[T] {
+	return &sizedElementsChannel[T]{
+		used: &atomic.Int64{},
+		cap:  int64(capacity),
+		ch:   make(chan T, capacity),
+	}
+}
+
+// enqueue puts the element into the queue with the given sized if there is enough capacity.
+// Returns an error if the queue is full. The callback is called before the element is committed to the queue.
+// If the callback returns an error, the element is not put into the queue and the error is returned.
+func (vcq *sizedElementsChannel[T]) enqueue(el T, size int64, callback func() error) error {
+	if vcq.initTotalSize+vcq.used.Add(size) > vcq.cap {
+		vcq.used.Add(^(size - 1))
+		return ErrQueueIsFull
+	}
+	if callback != nil {
+		if err := callback(); err != nil {
+			vcq.used.Add(^(size - 1))
+			return err
+		}
+	}
+	vcq.ch <- el
+	return nil
+}
+
+// initBulkEnqueue puts the elements into the queue with the given size. It's used by the persistent queue to
+// initialize the queue with the elements recovered from the disk. Cannot be called concurrently.
+// Callers must ensure that the queue is empty when calling this function.
+func (vcq *sizedElementsChannel[T]) initBulkEnqueue(els []T, totalSize int64) {
+	vcq.initElementsCount = int64(len(els))
+	vcq.initTotalSize = totalSize
+	if cap(vcq.ch) < len(els) {
+		vcq.ch = make(chan T, len(els))
+	}
+	for _, el := range els {
+		vcq.ch <- el
+	}
+}
+
+// dequeue removes the element from the queue and returns it.
+// The call blocks until there is an item available or the queue is stopped.
+// The function returns true when an item is consumed or false if the queue is stopped and emptied.
+// The callback is called before the element is removed from the queue. It must return the size of the element.
+func (vcq *sizedElementsChannel[T]) dequeue(callback func(T) (size int64)) (T, bool) {
+	el, ok := <-vcq.ch
+	if !ok {
+		return el, false
+	}
+
+	size := callback(el)
+
+	if vcq.initElementsCount > 0 {
+		vcq.initSizeMutex.Lock()
+		defer vcq.initSizeMutex.Unlock()
+		vcq.initElementsCount--
+		if vcq.initElementsCount == 0 || vcq.initTotalSize < size {
+			vcq.initTotalSize = 0
+		} else {
+			vcq.initTotalSize -= size
+		}
+		return el, true
+	}
+
+	vcq.used.Add(^(size - 1))
```

Comment:
https://pkg.go.dev/sync/atomic#Int64.Add you can do Add(-size) that is the reason I suggested to use `Int64` to simplify this.

Reply:
Oh, right. updated